flax/core/lift.py

```diff
@@ -720,7 +721,90 @@ def scanned(broadcast_vars, carry, scan_variable_groups, rng_groups, args):
       name='scan')
 
 
-def custom_vjp(fn: Callable[..., Any],
+C = TypeVar("C")
+def while_loop(cond_fn: Callable[[Scope, C], bool],
+               body_fn: Callable[[Scope, C], C],
+               scope: Scope, init: C,
+               carry_variables: CollectionFilter = False,
+               broadcast_variables: CollectionFilter = True,
+               split_rngs: Mapping[PRNGSequenceFilter, bool] = {}) -> C:
+  """Lifted version of jax.lax.while_loop.
+
+  The lifted scope is passed to `cond_fn` and `body_fn`.
+  Broadcasted variables are immutable. The carry variable are
+  mutable but cannot change shape and dtype.
+  This also means you cannot initialize variables inside
+  the body. Consider calling `body_fn` once manually before
+  calling `while_loop` if variable initialization is required.
+
+  Example::
+
+    def f(scope, x):
+      def cond_fn(scope, c):
+        return scope.get_variable('state', 'acc') < 10
+      def body_fn(scope, c):
+        acc = scope.variable('state', 'acc')
+        acc += 1
+        y = scope.child(nn.dense)(c, c.shape[-1])
+        return y
+
+      c = x
+      c = body_fn(scope, c)
+      return lift.while_loop(cond_fn, body_fn, scope, (), carry_variables='state')
+
+  Args:
+    body_fn: The body of the while loop.
+    cond_fn: Should return True as long as the loop should continue.
+    scope: The scope(s) which should be lifted into the loop.
+    init: The initial state passed to the loop
+    carry_variables: collections that are carried through the loop
+      and are therefore mutable (default: none).
+    broadcast_variables: collections that are closed over and are
+      therefore read-only (default: all collections)
+    split_rngs: Split PRNG sequences will be different for each loop iterations.
+      If split is False the PRNGs will be the same across iterations.
+  Returns:
+    The final state after executing the while loop. 
+  """
+  rng_groups, rng_splits = _unzip2(split_rngs.items())
+
+  def inner(scope_fn, repack_fn,
+            variable_groups, rng_groups, carry_init):
+    carry_variables, broadcast_variables = variable_groups
+
+    def make_loop_rngs(i):
+      local_rng_groups = []
+      for rng_group, rng_split in zip(rng_groups, rng_splits):
+        if rng_split:
+          rng_group = tree_map_rngs(lambda rng: random.fold_in(rng, -i), rng_group)
```

Comment:
why flip the sign?

Reply:
Otherwise cond_fn and body_fn share rngs.

Reply:
sure, I guess I was wondering why have the negative here rather than just below.  it's not important :)

Reply:
Oh I see you're right. This one is redundant